### ISSUE TITLE
Callbacks

### ISSUE THREAD

**Issue by LoganDark** [overlookmotel/lock-queue#9]
`await`?? Please?!

Callbacks _suck_.

**Comment by overlookmotel** [comment]
`.run()` and `.lock()` methods both return Promises, and the functions you provide to `.run()` and `.lock()` are expected to return a Promise. There are no callbacks involved.

So you can already do `await locker.run( ... )` and provide an `async` function.

Please read the README again and shout if you don't understand.

**Comment by LoganDark** [comment]
Well, if only you had documentation explaining that.

The only code you have in the README involves callbacks. It does say it returns a promise, but it says the Promise resolves with the result of the callback rather than the acquisition of the lock.

What I want to do is along the lines of this:

```ts
await locker.lock()

// ... code

await locker.unlock()
```

**Comment by LoganDark** [comment]
I read your code and it turns out this is impossible.

Like I said, callbacks. The callback is here:

```ts
await locker.lock(fn, ...)
                  ^^
```

**Comment by overlookmotel** [comment]
First off, can I just ask you to keep your tone a little more polite? I'm sure it's just your way of speaking but... probably like most people, I'm very busy, and as publisher of a load of modules on npm, responding to issues on Github and maintaining the modules etc is time-consuming, and cannot be my highest priority.

It's very annoying when people come in with a tone of "your module is rubbish/doesn't work right/the docs aren't clear" etc, as if they have a god-given right to the finest quality code given to them for free and changes they want should be made immediately!

Of course, I'm open to issues and PRs and anything that will improve the module. But please bare in mind that module authors offer their code to the community for free and with no return. So when you ask for help/explanation/changes, it's nice to have in mind that essentially you're asking a favour from someone you don't know, and act accordingly.

To be honest it's not motivating to give help when you feel like you're being shouted at.

Now, to your actual query...

The purpose of this module is to await a lock and then run some code. Once the code has run, the lock is released. So `fn` is the code that you want to run *within* the lock.

An alternative API without callbacks would be:

```js
const lock = await locker.lock();
await doMyThing();
lock.release();
```

However, then you have to remember to release the lock yourself. And what if there's an error in the `doMyThing()` function? You need to make sure you release the lock in that case. So then your code becomes:

```js
const lock = await locker.lock();

try {
  await doMyThing();
} finally {
  lock.release();
}
```

Personally, I think the current API is cleaner as it handles all this for you:

```js
await locker.lock( async function() {
  await doMyThing();
} );
```

In the above example, the lock is released when the callback completes, regardless of whether it errors or not.

What do you think?

**Comment by LoganDark** [comment]
I've basically resorted to this:

```ts
	/**
	 * Calls the function `method` (bound to `this`) with arguments `[obj, key,
	 * ...args`. This function makes sure the lock `id`/`key` is respected.
	 * `exclusive` defines whether the function runs on its own or in parallel
	 * with other operations on the same lock.
	 *
	 * @param {Function} method The method to call. It's called bound to `this`.
	 * @param {boolean} exclusive Whether the method runs on its own or parallel
	 * to others.
	 * @param {boolean} objectExclusive Whether NO OTHER METHODS can run on the
	 * same object until this method returns.
	 * @param obj The object to run the method on.
	 * @param {string | number | undefined} key The key to operate on. Leave
	 * this blank if you're operating on more than one key at once.
	 * @param args
	 * @returns {Promise<any>}
	 */
	protected async locked<T> (
		method : ( ... args : any[] ) => Promise<T>,
		objectExclusive : boolean,
		exclusive : boolean,
		obj : any,
		key : string | number | undefined,
		... args : any[]
	) : Promise<T> {
		const id = await Carbyne.resolveId ( obj )
		const path = id + ( key !== undefined ? '/' + key : '' )
		const objectLocker = await this.getLock ( id )
		const locker = await this.getLock ( path )

		if ( key ) {
			args.unshift ( key )
		}

		const lockerFunc = async () =>
			( exclusive ? locker.lock : locker.run ).bind ( locker ) (
				async () =>
					method.bind ( this ) (
						id,
						... args
					),
				this
			)

		if ( key !== undefined ) {
			return await ( objectExclusive ? objectLocker.lock : objectLocker.run ).bind ( objectLocker ) (
				lockerFunc,
				this
			)
		}

		return await lockerFunc ()
	}
```

And

```ts
	/**
	 * Gets the key `key` of object `obj` and returns it.
	 *
	 * This function uses a lock to prevent all other operations on the same key
	 * during its runtime. This prevents data corruption and is an overall safe
	 * practice to adopt.
	 *
	 * @param obj The object or ID to target.
	 * @param {string | number} key
	 * @returns {Promise<void>}
	 */
	async getKey (
		obj : any,
		key : string | number
	) : Promise<TCarbyneValue> {
		return await this.locked (
			this._getKey,
			false,
			true,
			obj,
			key
		)
	}
```

I have to have a separate method for every single method now. But nevermind. It works.

I haven't squeezed all the bugs out yet, but it passes (most of) the tests.
